File: docs/compare.js
export default class PersonalityComparator {
    constructor(names, attributes) {
        this.names = names;
        this.attributes = attributes;
        this.people = [];
        this.verbose = false;
    }

    beVerbose() {
        if (this.verbose) {
            console.log(...arguments);
        }
    }

    // Method to initialize people array
    initializePeople() {
        for (let index = 0; index < this.names.length; index++) {
            const name = this.names[index];
            const person = { name: name, attributes: [] };

            for (const category in this.attributes) {
                const attribute = { [category]: this.attributes[category][index] };
                person.attributes.push(attribute);
            }

            this.people.push(person);
            this.beVerbose(name, person);
        }
        this.beVerbose("Initialized People:", ...this.people);
    }

    // Method to calculate Euclidean distance
    euclideanDistance(vector1, vector2) {
        let result;
        if (vector1.length !== vector2.length) {
            throw new Error('Vectors must have the same length');
        }

        let sum = 0;
        for (let i = 0; i < vector1.length; i++) {
            sum += Math.pow(vector1[i] - vector2[i], 2);
        }

        result = Math.sqrt(sum);
        this.beVerbose(result);
        return result;
    }

    // Method to calculate similarity between two people
    calculateSimilarity(index1, index2) {
        this.beVerbose("First input:", index1, "Second input:", index2);
        const attributes1 = this.people[index1].attributes.map(attribute => Object.values(attribute)[0]);
        const attributes2 = this.people[index2].attributes.map(attribute => Object.values(attribute)[0]);
        const distance = this.euclideanDistance(attributes1, attributes2);
        return distance.toFixed(3);
    }

    // Method to sort people by similarity to a target person
    sortSimilar(targetPerson) {
        const similarityScores = [];
        console.log('this.people:', this.people);
        for (let person = 0; person < this.people.length; person++) {
            console.log('person:', person)
            // problem is coming from calculateSimilarity
        const similarity = this.calculateSimilarity(targetPerson, person)
            console.log('similarity:', similarity)
            console.log('person.name:', this.people[person].name);
            similarityScores.push([this.people[person].name, similarity]);
        }
        const sortedPeople = similarityScores.sort((a, b) => a[1] - b[1]);
        this.beVerbose(similarityScores, sortedPeople);
        return sortedPeople;
    }

    // Method to compare all people
    compareAllPeople() {
        const comparisons = [];

        for (let i = 0; i < this.people.length; i++) {
            for (let j = i + 1; j < this.people.length; j++) {
                const person1 = this.people[i];
                const person2 = this.people[j];

                const similarity = this.calculateSimilarity(i, j);
                const comparison = `${person1.name} vs ${person2.name}: ${similarity}`;
                comparisons.push([person1.name, person2.name, similarity]);
            }
        }
        this.beVerbose(comparisons);
        return comparisons;
    }
}

File: compare.py
import numpy as np
from sys import argv

names = ['Professor Parviz', 'Lalik', 'Allen', 'Hyun Yoon', 'Paloma', 'Jonah', 'Leana', 'Kylie', 'Siera', 'Truth', 'Fartun Aden', 'Dani Herd', 'Yana', 'Jocelyn', 'Satinder', 'Mya', 'Zahra', 'Osman Ali', 'Peter', 'Weihua', 'Enzo']

categories = {
    'openness': [5,3,4,2,4,4,5,5,5,5,5,3,3,4,3,4,3,3,4,2,3],
    'conscientiousness': [3,2,2,3,4,2,3,2,2,3,2,4,3,4,2,3,3,5,4,1,4],
    'extraversion': [3,4,1,2,5,4,5,0,5,3,4,2,0,2,1,5,3,5,2,4,5],
    'agreeableness': [5,5,4,2,4,5,5,4,2,4,4,4,5,5,3,5,4,3,5,2,2],
    'neuroticism': [5,3,3,1,4,2,3,3,3,3,1,2,2,2,2,5,2,1,2,3,1],
}

def get_people():
    people = []
    for index in range(0, len(names)):
        name = names[index]
        person = {'name': name, 'attributes': []}

        for category in categories:
            attribute = {category: categories[category][index]}
            person['attributes'].append(attribute)

        people.append(person)

    return people

def euclidean_distance(vector1, vector2):
    return np.linalg.norm(np.array(vector1) - np.array(vector2))

def calculate_similarity(person1, person2):
    attributes1 = [list(attribute.values())[0] for attribute in person1['attributes']]
    attributes2 = [list(attribute.values())[0] for attribute in person2['attributes']]
    distance = euclidean_distance(attributes1, attributes2)
    return distance

def sort_people_by_similarity(target_person, people_list):
    # Calculate similarity scores for the target person against others
    similarity_scores = [(person['name'], calculate_similarity(target_person, person)) for person in people_list]
    
    # Sort people by similarity (lower distance is more similar)
    sorted_people = sorted(similarity_scores, key=lambda x: x[1])

    return sorted_people

def compare_all_people(people_list):
    comparisons = []

    for i in range(len(people_list)):
        for j in range(i + 1, len(people_list)):
            person1 = people_list[i]
            person2 = people_list[j]

            similarity = calculate_similarity(person1, person2)
            comparison = f"{person1['name']} vs {person2['name']}: {similarity}"
            comparisons.append((person1['name'], person2['name'], similarity))

    return comparisons

def parse_args(args):
    options = ['-a', '-l']
    # we want to skip the first argument (the file name)
    for index in range(1, len(args)):
        line = args[index]
        options.append([{options[index]: True}] if line in ['-a', '--all'] else [{options[index]: False}])
        options.append([{options[index]: True}] if line in ['-l', '--list'] else [{options[index]: False}])

def main():
    options = parse_args(argv)
    people = get_people()

    for index in range(0, len(people)):
            print(people[index]['name']+':', index)
            # for person in people[index]:
                # print(person[0])

    if arg_flag:
        # Example: Compare every person to every other person
        all_comparisons = compare_all_people(people)

        # Display the comparisons
        for person1, person2, similarity in all_comparisons:
            print(f"{person1} vs {person2}: {similarity}")
    else:
        # Example: Sort people by similarity to the chosen person
        target = int(input("\nPick a person: \n"))
        target_person = people[target]
        sorted_people = sort_people_by_similarity(target_person, people)

        # Display the sorted list
        for person, distance in sorted_people:
            print(f"{distance}: {person}")

        # print('\nThe most similar person to '+str(target_person['name']), 'is', str(sorted_people[1][0]))

main()